### PR TITLE
Prevented unexpected exit: handle SIGPIPE properly

### PR DESCRIPTION
**Pull request by qianl15**
We found that CoreArbiter server sometimes crashed due to SIGPIPE. This signal often caused by writing to a broken TCP connection, which means the application that uses Arachne did not close the connection correctly (e.g., the application crashed without closing the connection). 

Thus, a temporary solution is to ignore `SIGPIPE` and the `sendData()` error, then the CoreArbiter server would eventually detect the closed connection and clean up the thread state.

### CODE REVIEW

File: src/CoreArbiterServer.cc [hunk 1]
```diff
@@ -1579,6 +1590,11 @@ CoreArbiterServer::installSignalHandler() {
         LOG(ERROR, "Couldn't set signal handler for SIGSEGV");
     if (sigaction(SIGABRT, &signalAction, NULL) != 0)
         LOG(ERROR, "Couldn't set signal handler for SIGABRT");
+
+    // Set sa_flags to 0 to ensure the function fail
```
Comment (hq6) [inline src/CoreArbiterServer.cc]: Why do want to "ensure the function fail?" Also, which function are we trying to ensure fails?

File: src/CoreArbiterServerMain.cc [hunk 2]
```diff
@@ -106,7 +106,7 @@ parseOptions(int* argcp, const char** argv) {
 }
 
 int main(int argc, const char** argv) {
-    Logger::setLogLevel(CoreArbiter::ERROR);
+    Logger::setLogLevel(CoreArbiter::NOTICE);
```
Comment (hq6) [inline src/CoreArbiterServerMain.cc]: This should probably be in its own commit. It is unrelated to other changes in this PR.

File: src/CoreArbiterServer.cc [hunk 3]
```diff
@@ -1067,8 +1067,12 @@ CoreArbiterServer::distributeCores()
                     if (!sendData(thread->socket, &core->id, sizeof(int),
                                   "Error sending core ID to thread " +
                                         std::to_string(thread->id))) {
-                        exit(-1);
-                        return;
+                        // TODO(Qian): is it a good solution to ignore error?
+                        // The client side did not close TCP connection 
+                        // in a normal way, which caused the failure
+                        LOG(WARNING, "Pending to clean up state for thread %s",
```
Comment (hq6) [inline src/CoreArbiterServer.cc]: I'm not sure this extra log message is adding anything, unless the original error on line 1068 is not printing for some reason?

File: src/CoreArbiterServer.cc [hunk 4]
```diff
@@ -1067,8 +1067,12 @@ CoreArbiterServer::distributeCores()
                     if (!sendData(thread->socket, &core->id, sizeof(int),
                                   "Error sending core ID to thread " +
                                         std::to_string(thread->id))) {
-                        exit(-1);
-                        return;
+                        // TODO(Qian): is it a good solution to ignore error?
```
Comment (hq6) [inline src/CoreArbiterServer.cc]: After thinking about this some more, I'd still be much more comfortable if we can do the cleanup here; otherwise there will be a window of vulnerability where the Core Arbiter Server incorrectly believes that a thread is assigned to a core (thus preventing that core from being used by any other thread) when the reality is that the thread is already dead.

**Review by hq6 — CHANGES_REQUESTED** [review]
See inline comments.

**Comment by qianl15** [comment]
Now we clean up the connection in place.

File: src/CoreArbiterServer.cc [hunk 5]
```diff
@@ -805,7 +805,10 @@ CoreArbiterServer::timeoutThreadPreemption(int timerFd)
 void
 CoreArbiterServer::cleanupConnection(int socket)
 {
-    sys->close(socket);
+    if (sys->close(socket) == -1) {
+        // The socket was closed before, just return
```
Comment (johnousterhout) [inline src/CoreArbiterServer.cc]: Aren't there many different errors that close can return? How do you know that it failed because the socket was already closed, as opposed to some other cause? Also, shouldn't you know whether you have already closed the socket?  Let's discuss.

File: src/CoreArbiterServer.cc [hunk 6]
```diff
@@ -805,7 +805,10 @@ CoreArbiterServer::timeoutThreadPreemption(int timerFd)
 void
 CoreArbiterServer::cleanupConnection(int socket)
```
Comment (johnousterhout) [inline src/CoreArbiterServer.cc:808]: I wonder if the argument to this method should be a ProcessInfo rather than an int. Let's discuss.

File: src/CoreArbiterServer.cc [hunk 7]
```diff
@@ -1067,8 +1070,12 @@ CoreArbiterServer::distributeCores()
                     if (!sendData(thread->socket, &core->id, sizeof(int),
                                   "Error sending core ID to thread " +
                                         std::to_string(thread->id))) {
-                        exit(-1);
-                        return;
+                        // The client side did not close TCP connection
```
Comment (johnousterhout) [inline src/CoreArbiterServer.cc]: How do you know this is the cause of the failure? Let's discuss.

File: src/CoreArbiterServer.cc [hunk 8]
```diff
@@ -1558,6 +1565,12 @@ void signalHandler(int signum) {
         }).detach();
     } else if ((signum == SIGSEGV) || (signum == SIGABRT)) {
         invokeGDB(signum);
+    } else if (signum == SIGPIPE) {
+        // Ignore SIGPIPE and allow repeated invocations
+        signalAction.sa_handler = signalHandler;
+        sigaction(signum, &signalAction, NULL);
+        LOG(WARNING, "Received SIGPIPE: connection was not closed normally");
```
Comment (johnousterhout) [inline src/CoreArbiterServer.cc]: Code in a signal handler is very dangerous: it's effectively unsynchronized with respect to the rest of the application (the signal can happen at any time, no?). Let's discuss whether this particular code is safe.  Also, I think there may be a way that you can stop SIGPIPE signals and just get errors on I/O operations instead. If so, I'd recommend using this approach.

File: src/CoreArbiterServer.cc [hunk 9]
```diff
@@ -1579,6 +1592,8 @@ CoreArbiterServer::installSignalHandler() {
         LOG(ERROR, "Couldn't set signal handler for SIGSEGV");
     if (sigaction(SIGABRT, &signalAction, NULL) != 0)
         LOG(ERROR, "Couldn't set signal handler for SIGABRT");
+    if (sigaction(SIGPIPE, &signalAction, NULL) != 0)
+        LOG(ERROR, "Couldn't set signal handler for SIGPIPE");
```
Comment (johnousterhout) [inline src/CoreArbiterServer.cc]: I'm not sure it makes sense to continue the application if you get one of these errors. Let's discuss.

File: src/CoreArbiterServerMain.cc [hunk 10]
```diff
@@ -118,6 +118,7 @@ int main(int argc, const char** argv) {
             printf(" %d", coresUsed[i]);
         putchar('\n');
     }
+    fflush(stdout);
```
Comment (johnousterhout) [inline src/CoreArbiterServerMain.cc:121]: Why is this needed? Let's discuss. Same with with line 67 below.

**Review by johnousterhout — COMMENTED** [review]
Let's get together to discuss this code review once you have read through it.

-John-

File: src/CoreArbiterServerMain.cc (hunk 10 above)
Comment (qianl15) [reply src/CoreArbiterServerMain.cc:121]: I think we need to standardize the logging: write to an error stream similar to Arachne rather than print to stdout.

File: src/CoreArbiterServer.cc (hunk 6 above)
Comment (qianl15) [reply src/CoreArbiterServer.cc:808]: We need to close the socket, clean up the thread, and update the process info. Socket fd here is also used to find the thread & process info. So I think it should be fine.

File: src/CoreArbiterServer.cc [hunk 11]
```diff
@@ -806,6 +806,10 @@ void
 CoreArbiterServer::cleanupConnection(int socket)
 {
     sys->close(socket);
```
Comment (qianl15) [inline src/CoreArbiterServer.cc]: LOG the error as a Warning here. Put the close before checking threadInfo.

File: src/CoreArbiterServer.cc [hunk 12]
```diff
@@ -806,6 +806,10 @@ void
...
+    if (threadSocketToInfo.find(socket) == threadSocketToInfo.end()) {
+        LOG(WARNING, "Unknown thread");
```
Comment (qianl15) [inline src/CoreArbiterServer.cc]: No need to log here.
